File: backend/orders/api/urls/adjustments.py
from __future__ import annotations

from django.urls import path

from orders.api.views.adjustments.client_counter_scope_views import (
    ClientCounterScopeIncrementView,
)
from orders.api.views.adjustments.extra_service_adjustment_views import (
    ClientAcceptExtraServiceView,
    CreateExtraServiceAdjustmentView,
)
from orders.api.views.adjustments.scope_increment_adjustment_views import (
    CreateScopeIncrementAdjustmentView,
)
from orders.api.views.adjustments.adjustment_negotiation_views import (
    AdjustmentAcceptView,
    AdjustmentCancelView,
    AdjustmentCounterView,
    AdjustmentCreateView,
    AdjustmentDeclineView,
    AdjustmentStaffOverrideView,
    ClientAcceptScopeRequestView,
    StaffResolveAdjustmentEscalationView,
    WriterEscalateAdjustmentView,
)
from orders.api.views.adjustments.adjustment_detail_views import (
    AdjustmentDetailView,
    LatestOrderAdjustmentView,
)
from orders.api.views.adjustments.adjustment_inbox_views import StaffAdjustmentInboxView

urlpatterns = [
    # --- Staff inbox ---
    path(
        "orders/adjustments/inbox/",
        StaffAdjustmentInboxView.as_view(),
        name="order-adjustment-inbox",
    ),

    # --- Create ---
    path(
        "orders/<int:order_id>/adjustments/",
        AdjustmentCreateView.as_view(),
        name="order-adjustment-create",
    ),
    path(
        "orders/<int:order_id>/adjustments/scope-increment/",
        CreateScopeIncrementAdjustmentView.as_view(),
        name="order-adjustment-scope-increment-create",
    ),
    path(
        "orders/<int:order_id>/adjustments/extra-service/",
        CreateExtraServiceAdjustmentView.as_view(),
        name="order-adjustment-extra-service-create",
    ),

    # --- Client actions ---
    path(
        "orders/adjustments/<int:adjustment_id>/counter-scope/",
        ClientCounterScopeIncrementView.as_view(),
        name="order-adjustment-counter-scope",
    ),
    path(
        "orders/adjustments/<int:adjustment_id>/accept-scope/",
        ClientAcceptScopeRequestView.as_view(),
        name="order-adjustment-accept-scope",
    ),
    path(
        "orders/adjustments/<int:adjustment_id>/accept-extra-service/",
        ClientAcceptExtraServiceView.as_view(),
        name="order-adjustment-accept-extra-service",
    ),
    path(
        "orders/adjustments/<int:adjustment_id>/counter/",
        AdjustmentCounterView.as_view(),
        name="order-adjustment-counter",
    ),
    path(
        "orders/adjustments/<int:adjustment_id>/accept/",
        AdjustmentAcceptView.as_view(),
        name="order-adjustment-accept",
    ),
    path(
        "orders/adjustments/<int:adjustment_id>/decline/",
        AdjustmentDeclineView.as_view(),
        name="order-adjustment-decline",
    ),
    path(
        "orders/adjustments/<int:adjustment_id>/cancel/",
        AdjustmentCancelView.as_view(),
        name="order-adjustment-cancel",
    ),

    # --- Writer escalation ---
    path(
        "orders/adjustments/<int:adjustment_id>/escalate/",
        WriterEscalateAdjustmentView.as_view(),
        name="order-adjustment-escalate",
    ),

    # --- Read ---
    path(
        "orders/adjustments/<int:adjustment_id>/",
        AdjustmentDetailView.as_view(),
        name="order-adjustment-detail",
    ),
    path(
        "orders/<int:order_id>/adjustments/latest/",
        LatestOrderAdjustmentView.as_view(),
        name="order-adjustment-latest",
    ),

    # --- Staff actions ---
    path(
        "orders/adjustments/<int:adjustment_id>/staff-override/",
        AdjustmentStaffOverrideView.as_view(),
        name="order-adjustment-staff-override",
    ),
    path(
        "orders/adjustments/<int:adjustment_id>/resolve-escalation/",
        StaffResolveAdjustmentEscalationView.as_view(),
        name="order-adjustment-resolve-escalation",
    ),
]

File: frontend/src/api/adjustments.ts
import { api, ordersApiPath } from "./client";
import type {
  AdjustmentRequest,
  CreateExtraServicePayload,
  CreateScopeIncrementPayload,
  StaffAdjustmentInboxResponse,
} from "@/types/adjustments";

const adj = (id: number | string) => ordersApiPath(`/orders/adjustments/${id}`);

export const adjustmentsApi = {
  // --- Read ---
  getDetail: (adjustmentId: number | string) =>
    api.get<AdjustmentRequest>(`${adj(adjustmentId)}/`),

  getLatest: (orderId: number | string) =>
    api.get<AdjustmentRequest>(ordersApiPath(`/orders/${orderId}/adjustments/latest/`)),

  inbox: (params?: Record<string, unknown>) =>
    api.get<StaffAdjustmentInboxResponse>(ordersApiPath("/orders/adjustments/inbox/"), { params }),

  // --- Create (writer / staff) ---
  createScopeIncrement: (orderId: number | string, payload: CreateScopeIncrementPayload) =>
    api.post<AdjustmentRequest>(ordersApiPath(`/orders/${orderId}/adjustments/scope-increment/`), payload),

  createExtraService: (orderId: number | string, payload: CreateExtraServicePayload) =>
    api.post<AdjustmentRequest>(ordersApiPath(`/orders/${orderId}/adjustments/extra-service/`), payload),

  // --- Client actions ---
  acceptScope: (adjustmentId: number | string) =>
    api.post<{ message: string; status: string }>(`${adj(adjustmentId)}/accept-scope/`, {}),

  counterScope: (adjustmentId: number | string, countered_quantity: number, countered_note = "") =>
    api.post<{ message: string; status: string }>(
      `${adj(adjustmentId)}/counter-scope/`,
      { countered_quantity, countered_note },
    ),

  acceptExtraService: (adjustmentId: number | string) =>
    api.post<{ message: string; status: string }>(`${adj(adjustmentId)}/accept-extra-service/`, {}),

  decline: (adjustmentId: number | string, reason: string) =>
    api.post<{ message: string; status: string }>(`${adj(adjustmentId)}/decline/`, { reason }),

  cancel: (adjustmentId: number | string, reason: string) =>
    api.post<{ message: string; status: string }>(`${adj(adjustmentId)}/cancel/`, { reason }),

  // --- Writer escalation ---
  escalate: (adjustmentId: number | string, reason: string) =>
    api.post<{ message: string }>(`${adj(adjustmentId)}/escalate/`, { reason }),

  // --- Deadline decrease / rush (client) ---
  createDeadlineDecrease: (
    orderId: number | string,
    payload: { new_deadline: string; reason?: string },
  ) =>
    api.post<{
      id: number;
      status: string;
      adjustment_type: string;
      new_deadline: string;
      surcharge: string;
      surcharge_breakdown: {
        original_multiplier: string;
        new_multiplier: string;
        original_hours: number;
        new_hours: number;
      };
      message: string;
    }>(
      ordersApiPath(`/orders/${orderId}/adjustments/deadline-decrease/`),
      payload,
    ),

  // --- Deadline extension (writer) ---
  createDeadlineExtension: (
    orderId: number | string,
    payload: { requested_deadline: string; reason: string; writer_justification?: string },
  ) =>
    api.post<{ id: number; status: string; requested_deadline: string; hours_requested: number | null }>(
      ordersApiPath(`/orders/${orderId}/adjustments/deadline-extension/`),
      payload,
    ),

  // --- Staff actions ---
  staffOverride: (adjustmentId: number | string, amount: string, notes: string) =>
    api.post<{ message: string; status: string }>(`${adj(adjustmentId)}/staff-override/`, { amount, notes }),

  resolveEscalation: (adjustmentId: number | string, resolution: string, note = "") =>
    api.post<{ message: string; status: string }>(`${adj(adjustmentId)}/resolve-escalation/`, { resolution, note }),
};
